staging/src/k8s.io/code-generator/cmd/validation-gen/output_tests/update_validations/primitives/zz_generated.validations.go

```diff
@@ -0,0 +1,100 @@
+//go:build !ignore_autogenerated
+// +build !ignore_autogenerated
+
+/*
+Copyright The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Code generated by validation-gen. DO NOT EDIT.
+
+package primitives
+
+import (
+	fmt "fmt"
+
+	validate "k8s.io/apimachinery/pkg/api/validate"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	field "k8s.io/apimachinery/pkg/util/validation/field"
+)
+
+func init() { localSchemeBuilder.Register(RegisterValidations) }
+
+// RegisterValidations adds validation functions to the given scheme.
+// Public to allow building arbitrary schemes.
+func RegisterValidations(scheme *runtime.Scheme) error {
+	scheme.AddValidationFunc((*T1)(nil), func(obj, oldObj interface{}, subresources ...string) field.ErrorList {
+		if len(subresources) == 0 {
+			return Validate_T1(obj.(*T1), oldObj.(*T1), oldObj != nil, nil)
+		}
+		return field.ErrorList{field.InternalError(nil, fmt.Errorf("no validation found for %T, subresources: %v", obj, subresources))}
+	})
+	return nil
+}
+
+func Validate_T1(obj *T1, oldObj *T1, isUpdate bool, fldPath *field.Path) (errs field.ErrorList) {
+	// field T1.S
+	var oldS *string = nil
+	if oldObj != nil {
+		oldS = &oldObj.S
+	}
+	errs = append(errs,
+		func(obj string, oldObj *string, fldPath *field.Path) (errs field.ErrorList) {
+			if isUpdate {
+				errs = append(errs, validate.FixedResultUpdate(fldPath, &obj, oldObj, true, "T1.S, UpdateOnly")...)
```

Comment:
Is it weird that one is a value and one is a pointer?  What if we made it always a pointer?  I'm not sure, I'd have to think that thru.  It feels weird though.

Reply:
This is really the crux of this design from a plugin perspective. 

For CEL for CRDs we started with accepting value and old value as the same type, but then later extended support to allow the old value to be optional (https://github.com/kubernetes/kubernetes/blob/5c14a5779551f678dbb89e9ce2346b9aa5e81d87/staging/src/k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/types_jsonschema.go#L216-L219). 

Note that we still don' call update directly on a value if a map entry is removed... so what we have is:

| value | old value | update validation call |
| - | - | - |
| absent | absent | skipped |
| absent | present | skipped |
| present | absent | called with nil pointer for old |
| present | present | called with non-nil pointer for old |

This is true even for PtrOK.

Should we support the whole table of options?  That would suggest that both value and oldValue be pointers.

We could also offer two options, one where the validator is called only if both values are present, another that is called for all cases?

Reply:
> ... That would suggest that both value and oldValue be pointers.

Yeah, that's what I was getting to.  It would change all of the validators to need to nil-check and deref, but it would be more consistent and maybe get rid of PtrOK flag?

I'll put it in TODO and think more about it.

Reply:
I just posted a few commits to make this change:

If a function has `UpdateOnly` and `PtrOK`, then it's always called with two pointers, but the oldValue pointer might be nil.

If a function has `UpdateOnly` (but does not have `PtrkOK`, it is only calls if oldValue is non-nil, and it's called with the non-pointer value of oldValue.

This seems like it might be kinda useful.  If you are writing a validation function that only cares about a change to an existing value, you don't have to deal with pointers, otherwise, you do.

Reply:
One question here:

Do we want this?

## UpdateOnly|PtrOK

| value | old value | update validation call |
| - | - | - |
| absent | absent | skipped |
| absent | present | called with pointers. value is nil |
| present | absent | called with pointers. oldValue is nil |
| present | present | called with non-nil pointers |

I have a local change that could implement this.